### ISSUE TITLE
Improving the NonEmpty experience

### ISSUE THREAD

**Issue by denizdogan** [dhall-lang/dhall-lang#1205]
I will be making pretty heavy usage of NonEmpty in my dhall-nginx project, and as it stands now, the end-user will be facing a lot of `NonEmpty.singleton`, and `NonEmpty.make` expressions. It doesn't feel unified unless the user decides to stick with a rather wordy `NonEmpty.make` expression every time, as in:

```dhall
NonEmpty.make Natural 3 ([] : List Natural)
```

Would it be possible to improve the experience? I'm not sure which options are realistic here, but two ideas stand out to me:

- Dedicated syntax for NonEmpty. All the good brackets are taken. Would something like `[]!` be a viable alternative?
- Add `NonEmpty.fromList xs` with some sort of validation like `assert : List/null xs === False`?

I'm sure others have put more thought into this than I have. :)

**Comment by SiriusStarr** [comment]
For packages at the least, I tend to just create builder functions, e.g.

```dhall
let reading = NonEmpty.singleton Reading.Type

let readings = NonEmpty.make Reading.Type

in  { reading, readings }
```
which then lets your actual data just have things like
```dhall
let R = ./Reading.dhall

in  { id = 21505, readings = R.readings "A reading" [ "Another reading" ] }
```

You could even go further by using records, e.g.

```dhall
let reading =
      \(r : Reading.Type) -> { readings = NonEmpty.singleton Reading.Type r }

let readings =
      \(r : Reading.Type) ->
      \(rs : Reading.Type) ->
        { readings = NonEmpty.make Reading.Type r rs }

in  { reading, readings }
```
and then use as
```dhall
let R = ./Reading.dhall

in  { id = 25105 } // R.readings "A reading" [ "Another reading" ]
```

Not saying it wouldn't potentially be nice to have a more ergonomic solution, but this pairs nicely with the fact that one is often using builder functions anyways.  Could make Haskell's `:|` syntactic sugar, maybe?  E.g.
```
{ readings = "A reading" :| [ "Another Reading" ] }
```
where `a :| b` desugars to `{head = a, tail = b}`?  But not sure how much nonemptys are going to be used for it to be worth adding more synactical constructs to the language.

**Comment by Gabriel439** [comment]
Another possible language-level change would be to require a type annotation for non-empty lists, but still use list syntax for the elements:

```dhall
[ 1, 2, 3 ] : NonEmpty Natural
```

**Comment by Gabriel439** [comment]
That said, I still think the validation-based approach using an `assert` will be the most ergonomic for the end user

**Comment by denizdogan** [comment]
What would an `assert`-based approach look like? I've tried my hands at validations before with limited success and understanding of them.

**Comment by Gabriel439** [comment]
@denizdogan: Yeah, it would basically be the same as the validation approach I mentioned in https://discourse.dhall-lang.org/t/config-validations/296

**Comment by MasseR** [comment]
Would this validation based approach be doable with just Dhall or does it require new built-in functions? I was trying to implement it with just Dhall and got this far:

``` dhall
\(TODO : forall (a : Type) -> a) ->
  let NonEmpty
      : forall (a : Type) -> Type
      = \(a : Type) -> { head : a, tail : List a }

  let proofOfNonEmpty =
        \(a : Type) ->
        \(x : List a) ->
          Natural/isZero (List/length a x) === False

  let QED = assert : True === True

  let fromList
      : forall (a : Type) ->
        forall (x : List a) ->
        proofOfNonEmpty a x ->
          NonEmpty a
      = \(a : Type) ->
        \(x : List a) ->
        \(proof : proofOfNonEmpty a x) ->
          TODO (NonEmpty a)

  in  { Type = NonEmpty, fromList, QED }
```

But if I'm not mistaken, there is no way to write the implementation of `fromList` because there is no partial pattern matching for the list. There is no way to get the head of the list, without involving optionals.

**Comment by Gabriel439** [comment]
I don't believe it's possible to implement a `fromList`-like function in Dhall.  You might be interested in this thread, though: https://discourse.dhall-lang.org/t/implementing-vector-in-dhall-or-should-we-add-dependent-sums/68

**Comment by winitzki** [comment]
"Proof of non-empty" needs to be not just length not equal to zero but the witness must be enough to deduce the first element.
